hypervisor/boot/reloc.c

```diff
@@ -73,10 +75,10 @@ void relocate(void)
 		for (dyn = (struct Elf64_Dyn *)_DYNAMIC; dyn->d_tag != DT_NULL; dyn++) {
 			switch (dyn->d_tag) {
 			case DT_RELA:
-				start = (struct Elf64_Rel *)(dyn->d_ptr + delta);
+				rela_start = (uint8_t *)(dyn->d_ptr + delta);
```

Comment:
Seems we don't need the new variable rela_start. Keep things simple. How about something like this:

@@ -95,7 +95,7 @@ void relocate(void)
                primary_32_start = (uint64_t)(&cpu_primary_start_32) - delta;
                primary_32_end = (uint64_t)(&cpu_primary_start_64) - delta;
 
-               while (start < end) {
+               while (rela_size >= size) {
                        if ((elf64_r_type(start->r_info)) == R_X86_64_RELATIVE) {
                                addr = (uint64_t *)(delta + start->r_offset);
 
@@ -120,6 +120,7 @@ void relocate(void)
                                }
                        }
                        start = (struct Elf64_Rel *)((char *)start + size);
+                       rela_size -= size;
                }
        }

BTW, with the new variable rela_size, to make it clearer,  probably we can change 'size' to 'entry_size'?

Reply:
Okay, I will change the 'size' to ' entry_size', Thanks